src/object/srv_obj.c

```diff
@@ -790,129 +797,152 @@ ds_obj_check_dtx_config(uint32_t flags, struct dtx_id *dti)
 	struct obj_tls			*tls = obj_tls_get();
 	daos_handle_t			 ioh = DAOS_HDL_INVAL;
 	uint32_t			 map_ver = 0;
 	uint32_t			 flags = 0;
-	int				 tag;
-	bool				 update;
-	bool				 dispatch;
 	bool				 resend = false;
-	int				 dti_cos_count = 0;
 	int				 rc;
-	int				 rc1;
 
 	D_ASSERT(orw != NULL);
 	D_ASSERT(orwo != NULL);
-	update = (opc_get(rpc->cr_opc) == DAOS_OBJ_RPC_UPDATE);
-	dispatch = update && orw->orw_shard_tgts.ca_arrays != NULL;
-	tag = dss_get_module_info()->dmi_tgt_id;
 
 	D_TIME_START(tls->ot_sp, OBJ_PF_UPDATE);
-	D_TIME_START(tls->ot_sp, OBJ_PF_UPDATE_PREP);
 
-	rc = ds_check_container(orw->orw_co_hdl, orw->orw_co_uuid,
-				&cont_hdl, &cont);
+	rc = ds_pre_check(orw->orw_map_ver, orw->orw_co_hdl, orw->orw_co_uuid,
+			  opc_get(rpc->cr_opc), &cont_hdl, &cont);
 	if (rc)
 		goto out;
 
-	if (update && !(cont_hdl->sch_capas & DAOS_COO_RW)) {
-		D_ERROR("cont "DF_UUID" sch_capas "DF_U64", "
-			"NO_PERM to update.\n",
-			DP_UUID(orw->orw_co_uuid), cont_hdl->sch_capas);
-		D_GOTO(out, rc = -DER_NO_PERM);
-	}
-
 	D_ASSERT(cont_hdl->sch_pool != NULL);
 	map_ver = cont_hdl->sch_pool->spc_map_version;
 
 	D_DEBUG(DB_TRACE, "rpc %p opc %d "DF_UOID" dkey %d %s tag/xs %d/%d eph "
 		DF_U64", pool ver %u/%u with "DF_DTI".\n",
 		rpc, opc_get(rpc->cr_opc), DP_UOID(orw->orw_oid),
 		(int)orw->orw_dkey.iov_len, (char *)orw->orw_dkey.iov_buf,
-		tag, dss_get_module_info()->dmi_xs_id, orw->orw_epoch,
+		dss_get_module_info()->dmi_tgt_id,
+		dss_get_module_info()->dmi_xs_id, orw->orw_epoch,
 		orw->orw_map_ver, map_ver, DP_DTI(&orw->orw_dti));
 
-	if (orw->orw_map_ver < map_ver) {
-		D_DEBUG(DB_IO, "stale version req %d map_version %d\n",
-			orw->orw_map_ver, map_ver);
-		if (update)
-			D_GOTO(out, rc = -DER_STALE);
-
-		/* It is harmless if fetch with old pool map version. */
-	} else if (orw->orw_map_ver > map_ver) {
-		/* XXX: Client (or leader replica) has newer pool map than
-		 *	current replica. Two possibile cases:
-		 *
-		 *	1. The current replica was the old leader if with
-		 *	   the old pool map version. According to current
-		 *	   leader election algorithm, it is still the new
-		 *	   leader with the new pool map version. Since no
-		 *	   leader switch, the unmatched pool version will
-		 *	   not affect DTX related availability check.
-		 *
-		 *	2. The current replica was NOT the old leader if
-		 *	   with the old pool map version. But it becomes
-		 *	   the new leader with the new pool map version.
-		 *	   In the subsequent modificaiton, it may hit
-		 *	   some 'prepared' DTX when make availability
-		 *	   check, it will return -DER_INPROGRESS that
-		 *	   will cause client to retry. It is possible
-		 *	   that the pool map version event arrives at
-		 *	   this server during the client retry. It is
-		 *	   inefficient, but harmless.
-		 */
-	}
-
-	if (!update) {
-		D_TIME_END(tls->ot_sp, OBJ_PF_UPDATE_PREP);
+	/** Three cases
+	 * 1. Single replica update.
+	 * 2. Non-leader update.
```

Comment:
non-leader also needs to check resend case.

Reply:
ah, resend check should be moved to ahead of this.